src/providers/files/files_ops.c

```diff
@@ -696,101 +684,287 @@ static errno_t sf_enum_groups(struct files_id_ctx *id_ctx,
         }
     }
 
-    ret = refresh_override_attrs(id_ctx, SYSDB_MEMBER_GROUP);
-    if (ret != EOK) {
-        DEBUG(SSSDBG_MINOR_FAILURE,
-              "Failed to refresh override attributes, "
-              "override values might not be available.\n");
+    if (groups[i] == NULL) {
+        ret = refresh_override_attrs(id_ctx, SYSDB_MEMBER_GROUP);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_MINOR_FAILURE,
+                  "Failed to refresh override attributes, "
+                  "override values might not be available.\n");
+        }
+
+        ret = EOK;
+    } else {
+        ret = EAGAIN;
     }
 
-    ret = EOK;
 done:
     talloc_free(tmp_ctx);
     return ret;
 }
 
-static errno_t sf_enum_files(struct files_id_ctx *id_ctx,
-                             uint8_t flags)
+enum update_steps {
+    DELETE_USERS,
+    READ_USERS,
+    SAVE_USERS,
+    DELETE_GROUPS,
+    READ_GROUPS,
+    SAVE_GROUPS,
+    UPDATE_FINISH,
+    UPDATE_DONE,
+};
+
+struct sf_enum_files_state {
+    struct files_id_ctx *id_ctx;
+    uint8_t flags;
+    struct tevent_timer *te;
+    enum update_steps current_step;
+    size_t step;
+    bool in_transaction;
+    size_t batch_size;
+    size_t obj_idx;
+    size_t file_idx;
+    struct passwd **users;
+    struct group **groups;
+    uint32_t delay;
+    uint32_t initial_delay;
+};
+
+static void sf_enum_files_first_step(struct tevent_context *ev,
+                                     struct tevent_timer *te,
+                                     struct timeval tv,
+                                     void *data);
+static struct tevent_req *sf_enum_files_send(struct files_id_ctx *id_ctx,
+                                             uint8_t flags)
 {
+    struct tevent_req *req;
+    struct sf_enum_files_state *state;
+    struct timeval tv;
     errno_t ret;
-    errno_t tret;
-    bool in_transaction = false;
+
+    req = tevent_req_create(id_ctx, &state, struct sf_enum_files_state);
+    if (req == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Unable to create tevent request!\n");
+        return NULL;
+    }
+
+    state->id_ctx = id_ctx;
+    state->flags = flags;
+    state->step = 0;
+    state->batch_size = 10;
+    state->obj_idx = 0;
+    state->file_idx = 0;
+    state->initial_delay = 100;
+    state->delay = 100;
+
+    if (state->flags & SF_UPDATE_PASSWD) {
+        state->current_step = DELETE_USERS;
+    } else if (state->flags & SF_UPDATE_GROUP) {
+        state->current_step = DELETE_GROUPS;
+    }
+
+    tv = tevent_timeval_current_ofs(0, state->initial_delay);
+    state->te = tevent_add_timer(id_ctx->be->ev, state, tv,
```

Comment:
This is triggered by signal that file has been modified. What in the case of scenario when file has been modified multiple times in a very short time span (t << state->initial_delay)? If we will not create multiple tevent chains running in parallel?

Reply:
I run small test:
1) Changed delay to be 100ms from 100us
2) Added some debug info every time sf_enum_files_first_step() us called
3) Assigned each batching random number to track it logs
4) Called `touch /etc/passwd & touch /etc/passwd & touch /etc/passwd & touch /etc/passwd &`

Result was creation of 3 almost parallel tracks of batching procedure:

```
./sssd_files.log:3558:(2021-04-01 16:30:17): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_USERS, track: 10
./sssd_files.log:3559:(2021-04-01 16:30:17): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_USERS, track: 5
./sssd_files.log:3560:(2021-04-01 16:30:17): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_USERS, track: 76
./sssd_files.log:3561:(2021-04-01 16:30:17): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: DELETE_GROUPS, track: 10
./sssd_files.log:3562:(2021-04-01 16:30:17): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: DELETE_GROUPS, track: 5
./sssd_files.log:3563:(2021-04-01 16:30:17): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: DELETE_GROUPS, track: 76
./sssd_files.log:3564:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_GROUPS, track: 10
./sssd_files.log:3620:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_GROUPS, track: 5
./sssd_files.log:3676:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_GROUPS, track: 76
./sssd_files.log:3732:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 10
./sssd_files.log:3809:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 5
./sssd_files.log:3832:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 76
./sssd_files.log:3855:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 10
./sssd_files.log:3939:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 5
./sssd_files.log:3963:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 76
./sssd_files.log:3987:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 10
./sssd_files.log:4071:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 5
./sssd_files.log:4095:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 76
./sssd_files.log:4119:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 10
./sssd_files.log:4203:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 5
./sssd_files.log:4227:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 76
./sssd_files.log:4251:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 10
./sssd_files.log:4335:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 5
./sssd_files.log:4359:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 76
./sssd_files.log:4383:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 10
./sssd_files.log:4428:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 5
./sssd_files.log:4443:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: SAVE_GROUPS, track: 76
./sssd_files.log:4458:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_GROUPS, track: 10
./sssd_files.log:4459:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_GROUPS, track: 5
./sssd_files.log:4460:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: READ_GROUPS, track: 76
./sssd_files.log:4461:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: UPDATE_FINISH, track: 10
./sssd_files.log:4489:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: UPDATE_FINISH, track: 5
./sssd_files.log:4517:(2021-04-01 16:30:18): [be[files]] [sf_enum_files_first_step] (0x0200): BATCH step: UPDATE_FINISH, track: 76
```

With the default 100us timeout there is a small chance it will happen using command line and bash scripts but it is still possible if someone will send inotify signal fast enough the other way.

I think every new batching routine may want to check if old routine is still running and use this knowledge somehow (canceling old routine as new will overwrite it, building on top of old routine by merging results etc.).

Reply:
The other interesting result from test above is that READ case is always fast, the SAVE case looks to be the slowest part. In case of frequent file changes it should be possible to run fast READ to get the diff of changes and then eventually inject it in already running SAVE routine from earlier batching.